"Sequence has more than one element" when a BindingContainer has multiple ContributesTo annotations

### Description

The reproducer throws the following exception:

```
Somewhere in file /private/var/folders/t2/qm93rrl178147bxlql0y3ljm0000gn/T/dev.zacsweers.metro.compiler.BoxTestGenerated$BindingcontainerstestMultipleContributesTo/kotlin-sources/main/MultipleContributesTo.kt: java.lang.IllegalArgumentException: Sequence has more than one element.
org.jetbrains.kotlin.util.FileAnalysisException: Somewhere in file /private/var/folders/t2/qm93rrl178147bxlql0y3ljm0000gn/T/dev.zacsweers.metro.compiler.BoxTestGenerated$BindingcontainerstestMultipleContributesTo/kotlin-sources/main/MultipleContributesTo.kt: java.lang.IllegalArgumentException: Sequence has more than one element.
	at org.jetbrains.kotlin.util.AnalysisExceptionsKt.wrapIntoFileAnalysisExceptionIfNeeded(AnalysisExceptions.kt:62)
...
Caused by: java.lang.IllegalArgumentException: Sequence has more than one element.
	at kotlin.sequences.SequencesKt___SequencesKt.single(_Sequences.kt:340)
	at dev.zacsweers.metro.compiler.fir.generators.ContributedInterfaceSupertypeGenerator.getScopedContributions(ContributedInterfaceSupertypeGenerator.kt:160)
	at dev.zacsweers.metro.compiler.fir.generators.ContributedInterfaceSupertypeGenerator.inCompilationScopesToContributions$lambda$3(ContributedInterfaceSupertypeGenerator.kt:102)
```

The workaround is to split this module as follows:

```kotlin
@BindingContainer
object BoxContainer {
  @Provides
  fun provideEncryptedBox(): Box {
    return RealBox(int = 1)
  }
}

@BindingContainer(includes = [BoxContainer::class])
@ContributesTo(InternalReleaseAppScope::class)
object InternalReleaseAppScopeBoxContainer

@BindingContainer(includes = [BoxContainer::class])
@ContributesTo(ProductionAppScope::class)
object ProductionAppScopeBoxContainer
```

But we've got lots of examples of this pattern and I'd love to avoid converting them all!

### Self-contained Reproducer

```kotlin
interface Box {
  val int: Int
}

class RealBox(override val int: Int) : Box

@Scope
annotation class AppScope

@Scope
annotation class InternalDebugAppScope

@Scope
annotation class InternalReleaseAppScope

@Scope
annotation class ProductionAppScope

@BindingContainer
@ContributesTo(InternalReleaseAppScope::class)
@ContributesTo(ProductionAppScope::class)
object BoxContainer {
  @Provides
  fun provideEncryptedBox(): Box {
    return RealBox(int = 1)
  }
}

@DependencyGraph(
  scope = AppScope::class,
  additionalScopes = [
    InternalReleaseAppScope::class,
  ],
)
interface VariantAppComponent {
  @EncryptedBox val encryptedBox: Box
}

fun box(): String {
  val graph = createGraph<VariantAppComponent>()
  assertEquals(graph.encryptedBox.int, 2)
  return "OK"
}
```

### Metro version

latest master

### Context

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,7 @@ Changelog
 - **Fix**: Check `rawStatus` for overrides when merging contributed supertypes.
 - **Fix**: Correctly extract the element type when creating implicit `Set` multibindings from `@ElementsIntoSet` contributors.
 - **Fix**: Check `additionalScopes` when merging binding containers too.
+- **Fix**: Don't fail if multiple contributing annotations on binding containers match the target scope when aggregating them.
 
 0.5.3
 -----

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -13,6 +13,7 @@ Changelog
 - **Fix**: Correctly extract the element type when creating implicit `Set` multibindings from `@ElementsIntoSet` contributors.
 - **Fix**: Check `additionalScopes` when merging binding containers too.
 - **Fix**: Don't fail if multiple contributing annotations on binding containers match the target scope when aggregating them.
+- **Fix**: Dedupe binding containers during graph node generation.
 
 0.5.3
 -----

--- a/compiler-tests/src/test/data/box/bindingcontainers/DedupeContainersContributedToMultipleScopes.kt
+++ b/compiler-tests/src/test/data/box/bindingcontainers/DedupeContainersContributedToMultipleScopes.kt
@@ -1,0 +1,38 @@
+// https://github.com/ZacSweers/metro/issues/861
+interface Box {
+  val int: Int
+}
+
+class RealBox(override val int: Int) : Box
+
+@Scope annotation class AppScope
+
+@Scope annotation class InternalDebugAppScope
+
+@Scope annotation class InternalReleaseAppScope
+
+@Scope annotation class ProductionAppScope
+
+@Qualifier annotation class EncryptedBox
+
+@BindingContainer
+@ContributesTo(InternalReleaseAppScope::class)
+@ContributesTo(ProductionAppScope::class)
+object BoxContainer {
+  @EncryptedBox
+  @Provides
+  fun provideEncryptedBox(): Box {
+    return RealBox(int = 1)
+  }
+}
+
+@DependencyGraph(scope = AppScope::class, additionalScopes = [InternalReleaseAppScope::class])
+interface VariantAppComponent {
+  @EncryptedBox val encryptedBox: Box
+}
+
+fun box(): String {
+  val graph = createGraph<VariantAppComponent>()
+  assertEquals(1, graph.encryptedBox.int)
+  return "OK"
+}

--- a/compiler-tests/src/test/java/dev/zacsweers/metro/compiler/BoxTestGenerated.java
+++ b/compiler-tests/src/test/java/dev/zacsweers/metro/compiler/BoxTestGenerated.java
@@ -144,6 +144,12 @@ public class BoxTestGenerated extends AbstractBoxTest {
     }
 
     @Test
+    @TestMetadata("DedupeContainersContributedToMultipleScopes.kt")
+    public void testDedupeContainersContributedToMultipleScopes() {
+      runTest("compiler-tests/src/test/data/box/bindingcontainers/DedupeContainersContributedToMultipleScopes.kt");
+    }
+
+    @Test
     @TestMetadata("MultibindsOnlyInContainer.kt")
     public void testMultibindsOnlyInContainer() {
       runTest("compiler-tests/src/test/data/box/bindingcontainers/MultibindsOnlyInContainer.kt");

--- a/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/generators/ContributedInterfaceSupertypeGenerator.kt
+++ b/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/generators/ContributedInterfaceSupertypeGenerator.kt
@@ -147,14 +147,11 @@ internal class ContributedInterfaceSupertypeGenerator(session: FirSession) :
     return buildMap {
       for (originClass in contributingClasses) {
         if (originClass.isAnnotatedWithAny(session, session.classIds.bindingContainerAnnotations)) {
-          val scopeId =
-            originClass
-              .annotationsIn(session, session.classIds.contributesToAnnotations)
-              .single()
-              .resolvedScopeClassId(typeResolver)
-          if (scopeId == scopeClassId) {
-            put(originClass.classId, true)
-          }
+          val hasMatchingScope =
+            originClass.annotationsIn(session, session.classIds.contributesToAnnotations).any {
+              it.resolvedScopeClassId(typeResolver) == scopeClassId
+            }
+          put(originClass.classId, hasMatchingScope)
           continue
         }
 

--- a/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/transformers/BindingContainerTransformer.kt
+++ b/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/transformers/BindingContainerTransformer.kt
@@ -706,6 +706,8 @@ internal class BindingContainer(
   val providerFactories: Map<CallableId, ProviderFactory>,
   val bindsMirror: BindsMirror?,
 ) {
+  private val classId = ir.classIdOrFail
+
   /**
    * Simple classes with a public, no-arg constructor can be managed directly by the consuming
    * graph.
@@ -714,4 +716,17 @@ internal class BindingContainer(
 
   fun isEmpty() =
     includes.isEmpty() && providerFactories.isEmpty() && (bindsMirror?.isEmpty() ?: true)
+
+  override fun equals(other: Any?): Boolean {
+    if (this === other) return true
+    if (javaClass != other?.javaClass) return false
+
+    other as BindingContainer
+
+    return classId == other.classId
+  }
+
+  override fun hashCode(): Int = classId.hashCode()
+
+  override fun toString(): String = classId.asString()
 }

